src/batch/src/executor2/generic_exchange.rs

```diff
@@ -128,60 +136,77 @@ impl<CS: 'static + CreateSource> Executor2 for GenericExchangeExecutor2<CS> {
-                self.current_source = Some(source);
+    async fn do_execute(self: Box<Self>) {
+        let mut sources: Vec<Box<dyn ExchangeSource>> = vec![];
+
+        for prost_source in &self.sources {
+            let source =
+                CS::create_source(self.env.clone(), prost_source, self.task_id.clone()).await?;
+            sources.push(source);
+        }
+
+        let mut handles = sources
+            .into_iter()
+            .map(data_chunk_stream)
+            .map(|stream| stream.into_future())
+            .collect_vec();
+
+        while !handles.is_empty() {
+            let ((result, from), _id, remainings) = select_all(handles).await;
+            handles = remainings;
+
+            if let Some(data_chunk) = result {
+                let data_chunk = data_chunk?;
+                handles.push(from.into_future());
+                yield data_chunk
             }
-            let mut source = self.current_source.take().unwrap();
-            match source.take_data().await? {
-                None => {
-                    self.current_source = None;
-                    self.source_idx += 1;
-                }
-                Some(res) => {
-                    if res.cardinality() == 0 {
-                        debug!("Exchange source {:?} output empty chunk.", source);
-                        assert_ne!(res.cardinality(), 0);
-                    }
-                    self.current_source = Some(source);
-                    yield res;
-                }
+        }
+    }
+}
+#[try_stream(boxed, ok = DataChunk, error = RwError)]
+async fn data_chunk_stream(mut source: Box<dyn ExchangeSource>) {
+    loop {
+        if let Some(res) = source.take_data().await? {
+            if res.cardinality() == 0 {
+                debug!("Exchange source {:?} output empty chunk.", source);
+                assert_ne!(res.cardinality(), 0);
             }
+            yield res;
+            continue;
         }
+        break;
     }
 }
-
 #[cfg(test)]
 mod tests {
     use std::sync::Arc;
 
     use futures::StreamExt;
+    use rand::Rng;
     use risingwave_common::array::column::Column;
     use risingwave_common::array::{DataChunk, I32Array};
     use risingwave_common::array_nonnull;
     use risingwave_common::types::DataType;
+    use tokio::time::{sleep, Duration};
 
     use super::*;
 
     #[tokio::test]
     async fn test_exchange_multiple_sources() {
-        #[derive(Debug)]
+        #[derive(Debug, Clone)]
         struct FakeExchangeSource {
-            chunk: Option<DataChunk>,
+            chunks: Vec<Option<DataChunk>>,
         }
 
         #[async_trait::async_trait]
         impl ExchangeSource for FakeExchangeSource {
             async fn take_data(&mut self) -> Result<Option<DataChunk>> {
-                let chunk = self.chunk.take();
-                Ok(chunk)
+                sleep(Duration::from_secs(0.1 as u64)).await;
```

Comment:
Do we have to sleep here?

Reply:
It's more convenient to assert data later.

Reply:
3 sources generate 3 same content data_chunk respectively
take example 
```
source 1 generate [1] [1] [1]
source 2 generate [2] [2] [2]
source 3 generate [3] [3] [3]
```
if not sleep for 0.1s the result order will be random 
else the result will look like that:
```
[1]
[3]
[2]

[3]
[2]
[1]

[2]
[1]
[3]
```

Reply:
`sleep` is probably not good for tests as we want to minimize the time spent on tests.
Maybe you can use conditional variables to trigger the returning of the data chunk.